wire-schema/src/main/java/com/squareup/wire/schema/SchemaLoader.java

```diff
@@ -29,118 +32,89 @@
-  public SchemaLoader addDirectory(File file) {
-    return addDirectory(file.toPath());
+  /** Add directory or zip file source from which proto files will be loaded. */
+  public SchemaLoader addSource(File file) {
+    return addSource(file.toPath());
   }
 
-  /** Add one or more directories from which proto files will be loaded. */
-  public SchemaLoader addDirectory(Path path) {
-    directories.add(path);
+  /** Add directory or zip file source from which proto files will be loaded. */
+  public SchemaLoader addSource(Path path) {
+    sources.add(path);
     return this;
   }
 
-  /** Returns a mutable list of the directories that this loader will load from. */
-  public List<Path> directories() {
-    return directories;
-  }
-
-  /**
-   * Add one or more proto files to load. Dependencies will be loaded automatically from the
-   * configured directories.
-   */
-  public SchemaLoader addProto(File file) {
-    return addProto(file.toPath());
+  /** Returns a mutable list of the sources that this loader will load from. */
+  public List<Path> sources() {
+    return sources;
   }
 
   /**
    * Add one or more proto files to load. Dependencies will be loaded automatically from the
-   * configured directories.
+   * configured sources.
    */
-  public SchemaLoader addProto(Path path) {
-    protos.add(path);
+  public SchemaLoader addProto(String proto) {
+    protos.add(proto);
     return this;
   }
 
   /** Returns a mutable list of the protos that this loader will load. */
-  public List<Path> protos() {
+  public List<String> protos() {
     return protos;
   }
 
   public Schema load() throws IOException {
-    if (directories.isEmpty()) {
-      throw new IllegalStateException("No directories added.");
+    if (sources.isEmpty()) {
+      throw new IllegalStateException("No sources added.");
     }
 
-    try {
-      // Attempt to load open each directory entry as a ZIP file. Most will fail; the rest are
-      // real ZIP files. We could ask first "is this a ZIP file" but trying and failing is easier!
-      for (Path directory : directories) {
-        if (Files.isRegularFile(directory)) {
-          try {
-            ZipFile zipFile = new ZipFile(directory.toFile());
-            zipFiles.put(directory, zipFile);
-          } catch (IOException ignored) {
-            // Not a ZIP file?!
-          }
-        }
-      }
-      return loadZipsAndDirectories();
-    } finally {
-      for (ZipFile zipFile : zipFiles.values()) {
-        try {
-          zipFile.close();
-        } catch (IOException ignored) {
+    try (Closer closer = Closer.create()) {
+      List<Path> directories = new ArrayList<>();
+      for (Path source : sources) {
+        if (Files.isRegularFile(source)) {
+          FileSystem sourceFs = FileSystems.newFileSystem(source, getClass().getClassLoader());
+          closer.register(sourceFs);
+          Iterables.addAll(directories, sourceFs.getRootDirectories());
```

Comment:
Yes. I'm not sure why they chose multiple, but every FS I've seen always
returns one.

On Sat, Sep 26, 2015, 10:12 AM Jesse Wilson <notifications@github.com>
wrote:

> In wire-schema/src/main/java/com/squareup/wire/schema/SchemaLoader.java
> <https://github.com/square/wire/pull/443#discussion_r40493932>:
>
> > -          }
> > -        }
> > -      }
> > -      return loadZipsAndDirectories();
> > -    } finally {
> > -      for (ZipFile zipFile : zipFiles.values()) {
> > -        try {
> > -          zipFile.close();
> > -        } catch (IOException ignored) {
> > +    try (Closer closer = Closer.create()) {
> > +      List<Path> directories = new ArrayList<>();
> > +      for (Path source : sources) {
> > +        if (Files.isRegularFile(source)) {
> > +          FileSystem sourceFs = FileSystems.newFileSystem(source, getClass().getClassLoader());
> > +          closer.register(sourceFs);
> > +          Iterables.addAll(directories, sourceFs.getRootDirectories());
>
> In practice this will have exactly one root, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/square/wire/pull/443/files#r40493932>.
>